### PR TITLE
Add possibility to proxy requests

### LINKED ISSUE #446: Support execution behind a corporate proxy
**User story.**
As a user working behind a corporate proxy, I can configure spectral, so that I can run it without errors when `ref$` points to internet based urls.

**Is your feature request related to a problem?**
In #245, @marbemac made it possible to resolve http based remote references. 

https://github.com/stoplightio/spectral/blob/027664fcf7f58f542a323d338b9017ba18adc9a3/src/resolvers/http.ts#L7

However, for those of us working behind a corporate proxy, well... things do tend to _not_ work. 

```
 842:13    error  [invalid-ref] FetchError: request to https://REDACTED/doco.yaml failed, reason: connect ETIMEDOUT XX.XX.XX.XX:443
```

**Describe the solution you'd like**
- Ideally, supporting `http_proxy` and `https_proxy` would just be awesome :tada:
- An optional argument to the cli would also fit my request.

**Additional context**
https://github.com/bitinn/node-fetch/issues/195

### PR DESCRIPTION
**Pull request by lag-of-death**
Fixes #446

**Checklist**

- [ ] Tests added / updated
- [x] Docs added / updated

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

## The solution

Instead of a global proxying mechanism, the requests made with `node-fetch`, were adjusted.
In the codebase, `node-fetch` is the only tool used for making requests and the changes in this PR adjust these requests to respect proxying.

We will need to remember about this in the future when possibly new requests are added. So, we will need to use `request.ts`
to handle proxying.

## Why proxy-agent?

1. Works for different node versions.
   
   This has the benefit of not making breaking changes that upgrading node to `>=10.0.0` would do (this would be needed with `global-agent`).

2. Supports http and https proxies.
3. There are only a couple of places where we make http calls, so there really is no need for a global proxying mechanism.

## Why NOT global-agent?

`global-agent` does not support HTTPS proxy. The docs say that it does, but it is not the case, `global-agent` only supports `HTTP` proxy. If we try and do sth like:

1. GLOBAL_AGENT_HTTPS_PROXY=https://127.0.0.1:8000 
2. GLOBAL_AGENT_HTTP_PROXY=https://127.0.0.1:8000 

then an exception is thrown: https://github.com/gajus/global-agent/blob/master/src/utilities/parseProxyUrl.js#L22

I haven't found any combination of `GLOBAL_AGENT_HTTPS_PROXY`/`GLOBAL_AGENT_HTTP_PROXY` and their values (urls to proxies - `http` vs `https`) to work with a proxy that's served on `HTTPS`.

The condition under that exception is thrown was added in v1.1.0: https://github.com/gajus/global-agent/releases/tag/v1.1.0
Haven't found a way around it (tried with env variables and setting `global.GLOBAL_AGENT.HTTPS_PROXY`).

##  Why NOT global-tunnel-ng

Will only work with node versions lower than 11.0.0.

### What was tested

| What was proxied                                 | http over http| http over https  | https over http   | https over https |
| ------------------------------------------------ |:-------------:|:----------------:|:-----------------:|:----------------:|
| --ruleset (the spec didn't have $ref)            | TESTED        | TESTED           | TESTED            | TESTED           |
| $ref from a spec (no `--ruleset http(s)://...` ) | TESTED        | TESTED           | TESTED            | TESTED           |

Also, various combinations of `--ruleset http(s)://...`  and `$ref: http(s)://...` were tested.
The testing was performed against the binary and `./dist/cli/index.js` as well.

#### What is what in the examples down below

1. http://localhost:3333 - http proxy 
2. https://localhost:4444 - https proxy 
3. http://localhost:9999 - a ruleset served with HTTP
4. https://localhost:7777 - a ruleset served with HTTPS

### Examples of the testing commands performed:

`NODE_TLS_REJECT_UNAUTHORIZED=0` was used because I was using self-signed certs.

1. `NODE_TLS_REJECT_UNAUTHORIZED=0 PROXY=http://localhost:3333 node ./dist/cli/index.js lint --ruleset http://localhost:9999 test_ref.yaml`
2. `NODE_TLS_REJECT_UNAUTHORIZED=0 PROXY=http://localhost:3333 node ./dist/cli/index.js lint --ruleset https://localhost:7777 test_ref.yaml`
3. `NODE_TLS_REJECT_UNAUTHORIZED=0 PROXY=https://localhost:4444 node ./dist/cli/index.js lint --ruleset https://localhost:7777 test_http_ref.yaml`
4. `NODE_TLS_REJECT_UNAUTHORIZED=0 PROXY=https://localhost:4444 node ./dist/cli/index.js lint --ruleset https://localhost:7777 test_https_ref.yaml`
5. `NODE_TLS_REJECT_UNAUTHORIZED=0 PROXY=https://localhost:4444 ./binaries/spectral-linux lint --ruleset https://localhost:7777 test_https_ref.yaml`
6. `NODE_TLS_REJECT_UNAUTHORIZED=0 PROXY=https://localhost:4444 ./binaries/spectral-linux lint --ruleset http://localhost:9999 test_https_ref.yaml`
7. `NODE_TLS_REJECT_UNAUTHORIZED=0 PROXY=https://localhost:4444 ./binaries/spectral-linux lint --ruleset http://localhost:9999 test_http_ref.yaml`
8. `NODE_TLS_REJECT_UNAUTHORIZED=0 PROXY=http://localhost:3333 ./binaries/spectral-linux lint test_http_ref.yaml`
9. `NODE_TLS_REJECT_UNAUTHORIZED=0 PROXY=http://localhost:3333 ./binaries/spectral-linux lint test_https_ref.yaml`
10. `NODE_TLS_REJECT_UNAUTHORIZED=0 PROXY=https://localhost:4444 ./binaries/spectral-linux lint test_https_ref.yaml`
11. `NODE_TLS_REJECT_UNAUTHORIZED=0 ./binaries/spectral-linux lint test_https_ref.yaml`
12. `./binaries/spectral-linux lint test_http_ref.yaml`
13. `./binaries/spectral-linux lint --ruleset http://localhost:9999 test_http_ref.yaml`

### Specs used for testing:

`test_ref.yaml`:
```yaml
openapi: 3.0.2
paths:
  /pets:
    get:
      responses:
        '200':
          description: abc
          content:
            application/json:
              schema:
                type: string
              example:
                123
```

`test_http_ref.yaml`:
```yaml
openapi: 3.0.2
paths:
  /pets:
    get:
      responses:
        '200':
          description: abc
          content:
            application/json:
              schema:
                $ref: http://localhost:8089/ok.json
              example:
                123
```

`http://localhost:8088/ok.json` was returning:
```json
"{"type":"string"}"
```
in order to get `"application/json.example" property type should be string`

`test_https_ref.yaml`:
```yaml
openapi: 3.0.2
paths:
  /pets:
    get:
      responses:
        '200':
          description: abc
          content:
            application/json:
              schema:
                $ref: https://localhost:8022/ok.json
              example:
                123
```

`https://localhost:8022/ok.json` was returning:
```json
"{"type":"string"}"
```
in order to get `"application/json.example" property type should be string`

### CODE REVIEW

**Comment by lag-of-death** [comment]
@philsturgeon, @marbemac, could you guys take a look at this PR and its description? Unsure what the next step should be here. `global-agent` seems to not be supporting https proxies and `global-tunnel-ng` prevents us from using node in versions >11.0.0 and I haven't been able to test it fully.

**Comment by philsturgeon** [comment]
This looks great! @nulltoken can you check this out? It looks like we might not need a v5.0 for this after-all, so we can tag a 4.2.0-alpha1 for you once this is confirmed as solving your use case.

**Comment by nulltoken** [comment]
@philsturgeon Awesome! I'll give it a try before the end of the week and will let you know about my findings.

File: README.md [hunk 1]
```diff
@@ -42,6 +42,12 @@ After [installation](https://stoplight.io/p/docs/gh/stoplightio/spectral/docs/ge
 - [Using the JavaScript API](https://stoplight.io/p/docs/gh/stoplightio/spectral/docs/guides/javascript.md)
 - [Creating custom functions](https://stoplight.io/p/docs/gh/stoplightio/spectral/docs/guides/custom-functions.md)
 
+## Proxying
```
Comment (philsturgeon) [inline README.md]: Could we put this in the CLI docs instead of the README?

File: src/request.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// @ts-ignore
+import * as fetch from 'node-fetch';
+import * as ProxyAgent from 'proxy-agent';
+
+export default (uri: string, opts: any = {}) => {
+  const options = process.env.PROXY ? { ...opts, agent: new ProxyAgent(process.env.PROXY) } : opts;
```
Comment (P0lip) [inline src/request.ts]: `process.env.PROXY` may throw a reference error in certain environments.

```suggestion
  const options = typeof process === 'object' && process.env.PROXY ? { ...opts, agent: new ProxyAgent(process.env.PROXY) } : opts;
```

File: src/request.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+// @ts-ignore
+import * as fetch from 'node-fetch';
+import * as ProxyAgent from 'proxy-agent';
+
+export default (uri: string, opts: any = {}) => {
```
Comment (P0lip) [inline src/request.ts]: Could add some typings if possible? :D

File: package.json [hunk 4]
```diff
@@ -59,6 +59,7 @@
     "@stoplight/path": "^1.2.0",
     "@stoplight/types": "^11.0.0",
     "@stoplight/yaml": "^3.1.0",
+    "@types/node-fetch": "^2.5.0",
```
Comment (P0lip) [inline package.json]: Could be a dev dep.

Comment (lag-of-death) [reply package.json]: thanks @P0lip, I'll look into adding the test

**Comment by lag-of-death** [comment]
hi @P0lip, can you take a look at the added tests ( https://github.com/stoplightio/spectral/blob/f8ecff2a1c0d458d29c1fc82216aa9205cd07558/src/__tests__/request.jest.test.ts ) ? couldn't have used `nock` here since `nock` only mocks requests, it couldn't be used for mocking a server. There are no mocks in the tests, but I think they clearly show what was the intent

**Comment by nulltoken** [comment]
> @philsturgeon Awesome! I'll give it a try before the end of the week and will let you know about my findings.

Sorry for the delay. Works amazingly!

@lag-of-death Tons of thanks for having made this possible! :heart_eyes:

**Comment by lag-of-death** [comment]
hi @nulltoken, no worries :) glad to be of help

**Comment by lag-of-death** [comment]
@P0lip, @philsturgeon, can anyone of you approve?

**Comment by philsturgeon** [comment]
Great job on this one!

**Pull request merged**